src/peergos/shared/hamt/Champ.java

```diff
@@ -0,0 +1,524 @@
+
+        System.arraycopy(src, 0, dst, 0, src.length);
+        dst[setIndex] = new HashPrefixPayload(null, MaybeMultihash.of(node.right));
+
+        return new Champ(dataMap, nodeMap, dst);
+    }
+
+    public CompletableFuture<Pair<Champ, Multihash>> remove(SigningPrivateKeyAndPublicHash writer,
+                                                            ByteArrayWrapper key,
+                                                            byte[] hash,
+                                                            int depth,
+                                                            MaybeMultihash expected,
+                                                            int bitWidth,
+                                                            int maxCollisions,
+                                                            ContentAddressedStorage storage,
+                                                            Multihash ourHash) {
+        int bitpos = mask(hash, depth, bitWidth);
+
+        if (dataMap.get(bitpos)) { // in place value
+            final int dataIndex = index(dataMap, bitpos);
+
+            HashPrefixPayload payload = contents[dataIndex];
+            KeyElement[] mappings = payload.mappings;
+            for (int payloadIndex = 0; payloadIndex < mappings.length; payloadIndex++) {
+                KeyElement mapping = mappings[payloadIndex];
+                final ByteArrayWrapper currentKey = mapping.key;
+                final MaybeMultihash currentVal = mapping.valueHash;
+                if (Objects.equals(currentKey, key)) {
+                    if (!currentVal.equals(expected)) {
+                        CompletableFuture<Pair<Champ, Multihash>> err = new CompletableFuture<>();
+                        err.completeExceptionally(new MutableTree.CasException(currentVal, expected));
+                        return err;
+                    }
+
+                    if (this.keyCount() == maxCollisions + 1 && this.nodeCount() == 0) {
+                    /*
+						 * Create new node with remaining pairs. The new node
+						 * will a) either become the new root returned, or b)
+						 * unwrapped and inlined during returning.
+						 */
+                        final BitSet newDataMap = BitSet.valueOf((depth == 0) ? dataMap.toByteArray() : new byte[0]);
+                        boolean lastInPrefix = mappings.length == 1;
+                        if (depth == 0 && lastInPrefix)
+                            newDataMap.clear(bitpos);
+                        else
+                            newDataMap.set(mask(hash, 0, bitWidth));
+
+                        HashPrefixPayload[] src = this.contents;
+                        HashPrefixPayload[] dst = new HashPrefixPayload[src.length - (lastInPrefix ? 1 : 0)];
+                        System.arraycopy(src, 0, dst, 0, dataIndex);
+                        System.arraycopy(src, dataIndex + 1, dst, dataIndex + (lastInPrefix ? 0 : 1), src.length - dataIndex - 1);
```

Comment:
The general pattern used in several places:
1. copy contents
2. do-something-with-node-at-some-index

Could this be refactored into something that can reduce the line  count?